add multi_student_t_cholesky

spinkney opened the pull request:
## Summary

Adds the Cholesky parameterization of the multivariate Student's T distribution. This PR uses autodiff and I'm planning a separate PR for the reverse mode adjoints.

## Release notes

The Cholesky parameterization of the multivariate Student's T distribution.

## Checklist

- [x] Math issue #2537 

- [x] Copyright holder: Sean Pinkney

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

bob-carpenter commented on test/unit/math/prim/prob/multi_student_t_cholesky_test.cpp:
```diff
@@ -0,0 +1,461 @@
+
+  double chi = 0;
+  for (int j = 0; j < K; j++)
+    chi += ((bin[j] - expect[j]) * (bin[j] - expect[j]) / expect[j]);
```
[optional]

If `bin` and `expect` are declared as Eigen vectors (and set to `Constant` to initialize), this can be done with vector operations, but it's messy, so I wouldn't recommend.  On the other hand, it makes it easy to at least compute `bin - expect` once.  But no big deal inside these tests.  

The empirical statistic is conventionally called `X`, not `chi` (don't ask me---I didn't invent stats' whacky notation).

spinkney replied:
done

bob-carpenter commented on test/unit/math/prim/prob/multi_student_t_cholesky_test.cpp:
```diff
@@ -0,0 +1,461 @@
+               std::invalid_argument);
+  EXPECT_THROW(stan::math::multi_student_t_cholesky_lpdf(y_3_2, nu, mu_3_3, L),
+               std::invalid_argument);
+}
```
We also need tests of boundary conditions like size 0 inputs.  Did I miss those?  You basically want to try to test all the branch points that return in the code.

spinkney replied:
These are all checked. Ctrl + F all these instances `std::numeric_limits<double>::quiet_NaN();`

spinkney commented:
@bob-carpenter thanks for the review! Note that I copied the current `multi_student_t_lpdf` code and left whatever was there except for the the changes necessary. I can correct all these and then I can go back through the `multi_student_t` and correct those. 

I'm going through and responding to each comment now.

spinkney commented on stan/math/prim/prob/multi_student_t_cholesky_lpdf.hpp:
```diff
@@ -0,0 +1,149 @@
+  check_not_nan(function, "Degrees of freedom parameter", nu);
+  check_positive(function, "Degrees of freedom parameter", nu);
+
+  if (is_inf(nu)) {
```
This is left over from the non-cholesky version. I think we should accept this current form because it follows the same standard as the non-cholesky version and then create a separate issue to remove this in both.

spinkney replied:
I take that back, I'm removing. If I use `inv_gamma_rng` it does not accept an `inf` (however boost's gamma does). So this should be consistent with our internal `inv_gamma_rng` function.

spinkney commented on stan/math/prim/prob/multi_student_t_cholesky_lpdf.hpp:
```diff
@@ -0,0 +1,149 @@
+                       "Size of another vector of "
+                       "the location variable",
+                       size_mu_old);
+      size_mu_old = size_mu_new;
```
cool! Should I also do the same for `y_vec` (the check before mu)?

bob-carpenter replied:
Yes, I think it's cleaner this way.

spinkney replied:
done

spinkney commented on stan/math/prim/prob/multi_student_t_cholesky_lpdf.hpp:
```diff
@@ -0,0 +1,149 @@
+    return multi_normal_cholesky_lpdf(y, mu, L);
+  }
+
+  size_t number_of_y = size_mvt(y);
```
done

spinkney commented on stan/math/prim/prob/multi_student_t_cholesky_lpdf.hpp:
```diff
@@ -0,0 +1,149 @@
+
+  vector_seq_view<T_y> y_vec(y);
+  vector_seq_view<T_loc> mu_vec(mu);
+  size_t size_vec = max_size_mvt(y, mu);
```
done

spinkney commented on stan/math/prim/prob/multi_student_t_cholesky_lpdf.hpp:
```diff
@@ -0,0 +1,149 @@
+
+  check_cholesky_factor(function, "scale parameter", L);
+
+  if (size_y == 0) {
```
done

spinkney commented on stan/math/prim/prob/multi_student_t_cholesky_lpdf.hpp:
```diff
@@ -0,0 +1,149 @@
+    lp -= (0.5 * size_y) * LOG_PI * size_vec;
+  }
+
+  using Eigen::Array;
```
done

spinkney commented on stan/math/prim/prob/multi_student_t_cholesky_rng.hpp:
```diff
@@ -0,0 +1,94 @@
+ * @param mu (Sequence of) location parameter(s)
+ * @param L Cholesky factor of the covariance matrix
+ * @param rng random number generator
+ * @throw std::domain_error if S is not positive definite, any value in mu is
```
done

spinkney commented on stan/math/prim/prob/multi_student_t_cholesky_rng.hpp:
```diff
@@ -0,0 +1,94 @@
+ * mu can be either an Eigen::VectorXd, an Eigen::RowVectorXd, or a std::vector
+ * of either of those types.
+ *
+ * @tparam T_loc Type of location parameter
```
done

spinkney commented on stan/math/prim/prob/multi_student_t_cholesky_rng.hpp:
```diff
@@ -0,0 +1,94 @@
+ * @tparam RNG Type of pseudo-random number generator
+ * @param nu degrees of freedom parameter
+ * @param mu (Sequence of) location parameter(s)
+ * @param L Cholesky factor of the covariance matrix
```
done

spinkney commented on stan/math/prim/prob/multi_student_t_cholesky_rng.hpp:
```diff
@@ -0,0 +1,94 @@
+  static const char* function = "multi_student_t_cholesky_rng";
+  check_not_nan(function, "Degrees of freedom parameter", nu);
+  check_positive(function, "Degrees of freedom parameter", nu);
+  check_positive(function, "Covariance matrix rows", L.rows());
```
done

spinkney commented on stan/math/prim/prob/multi_student_t_cholesky_rng.hpp:
```diff
@@ -0,0 +1,94 @@
+
+/** \ingroup multivar_dists
+ * Return a multivariate student-t random variate with the given degrees of
+ * freedom location and covariance using the specified random number generator.
```
done

spinkney commented on stan/math/prim/prob/multi_student_t_cholesky_rng.hpp:
```diff
@@ -0,0 +1,94 @@
+inline typename StdVectorBuilder<true, Eigen::VectorXd, T_loc>::type
+multi_student_t_cholesky_rng(
+    double nu, const T_loc& mu,
+    const Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic>& L, RNG& rng) {
```
done. I'd do the optional part if I understood how to.

spinkney commented on stan/math/prim/prob/multi_student_t_cholesky_rng.hpp:
```diff
@@ -0,0 +1,94 @@
+
+  variate_generator<RNG&, normal_distribution<> > std_normal_rng(
+      rng, normal_distribution<>(0, 1));
+  variate_generator<RNG&, gamma_distribution<> > gamma_rng(
```
done

spinkney commented on stan/math/prim/prob/multi_student_t_cholesky_rng.hpp:
```diff
@@ -0,0 +1,94 @@
+  for (size_t n = 0; n < N; ++n) {
+    Eigen::VectorXd z(L.cols());
+    for (int i = 0; i < L.cols(); i++) {
+      z(i) = std::sqrt(w) * std_normal_rng();
```
done

spinkney commented on stan/math/prim/prob/multi_student_t_cholesky_rng.hpp:
```diff
@@ -0,0 +1,94 @@
+ *    number of rows and columns in S
+ */
+template <typename T_loc, class RNG>
+inline typename StdVectorBuilder<true, Eigen::VectorXd, T_loc>::type
```
I don't know. I can update with something better, let me know.

bob-carpenter replied:
Ack---I meant to remove that comment after figuring out the polymorphism that was represented in the builder.  So please ignore this comment.

spinkney commented on stan/math/prim/prob/multi_student_t_cholesky_rng.hpp:
```diff
@@ -0,0 +1,94 @@
+template <typename T_loc, class RNG>
+inline typename StdVectorBuilder<true, Eigen::VectorXd, T_loc>::type
+multi_student_t_cholesky_rng(
+    double nu, const T_loc& mu,
```
interesting, I'll make this a separate issue

bob-carpenter replied:
thanks for making an issue so we don't forget.

spinkney commented on test/unit/math/prim/prob/multi_student_t_cholesky_test.cpp:
```diff
@@ -0,0 +1,461 @@
+  double lp = multi_student_t_cholesky_lpdf(y, nu, mu, L);
+  EXPECT_NO_THROW(multi_student_t_cholesky_rng(nu, mu, L, rng));
+  // calc using R's mnormt package's dmt function
+  EXPECT_NEAR(-10.1246, lp, 0.0001);
```
I'll see what precision I can get tomorrow.

spinkney replied:
I updated to 10 digits

bob-carpenter commented on test/unit/math/prim/prob/multi_student_t_cholesky_test.cpp:
```diff
@@ -0,0 +1,461 @@
+
+  int count = 0;
+  int bin[K];
+  double expect[K];
```
Anyway, no big deal if you can't get this changed.

spinkney replied:
I was able to get this in. The vector of int are initialized at 0 already and the `expect` needed to be size `K` 

```c++
 std::vector<int> bin(K);
 std::vector<double> expect(K, N / K);
```

spinkney commented on stan/math/prim/prob/multi_student_t_cholesky_lpdf.hpp:
```diff
@@ -0,0 +1,149 @@
+  vector_seq_view<T_loc> mu_vec(mu);
+  size_t size_vec = max_size_mvt(y, mu);
+
+  int size_y = y_vec[0].size();
```
done

spinkney commented:
@bob-carpenter I believe I addressed all the comments. I added the rev tests and I updated the `multi_student_t` code as well. It is ready for another review.

spinkney commented:
@rok-cesnovar jenkins says that it may be a configuration error?

bob-carpenter commented:
Thanks, @spinkney.  I'm traveling now, but I can review again on Monday.

spinkney commented:
@bob-carpenter this is ready for another round.
